File: api/app/schemas/stats.py
"""Schemas for stats"""

from datetime import date

from pydantic import BaseModel, Field


class ResidentFeeSumByMonthSchema(BaseModel):
    year: int
    month: int
    amounts: dict[str, float]
    total_usd: float


class EntityTransactionsByDaySchema(BaseModel):
    day: date
    transaction_count: int


class EntityMoneyFlowByDaySchema(BaseModel):
    day: date
    incoming_total_usd: float
    outgoing_total_usd: float


class TransactionsSumByWeekSchema(BaseModel):
    year: int
    week: int
    amounts: dict[str, float]
    total_usd: float


class EntityBalanceChangeByDaySchema(BaseModel):
    day: date
    balance_changes: dict[str, float]
    total_usd: float


class TransactionsSumByTagByMonthSchema(BaseModel):
    year: int
    month: int
    amounts: dict[str, float]
    total_usd: float


class TopEntityStatSchema(BaseModel):
    entity_id: int
    entity_name: str
    amounts: dict[str, float]
    total_usd: float


class TopTagStatSchema(BaseModel):
    tag_id: int
    tag_name: str
    amounts: dict[str, float]
    total_usd: float


class EntityStatsBundleSchema(BaseModel):
    cached: bool = True
    balance_changes: list[EntityBalanceChangeByDaySchema] = Field(default_factory=list)
    transactions_by_day: list[EntityTransactionsByDaySchema] = Field(
        default_factory=list
    )
    money_flow_by_day: list[EntityMoneyFlowByDaySchema] = Field(default_factory=list)
    top_incoming: list[TopEntityStatSchema] = Field(default_factory=list)
    top_outgoing: list[TopEntityStatSchema] = Field(default_factory=list)
    top_incoming_tags: list[TopTagStatSchema] = Field(default_factory=list)
    top_outgoing_tags: list[TopTagStatSchema] = Field(default_factory=list)

File: ui-new/src/api/stats.ts
import { apiRequest } from './client';

export type EntityBalanceChangeByDay = {
  day: string;
  balance_changes: Record<string, number>;
  total_usd: number;
};

export type EntityMoneyFlowByDay = {
  day: string;
  incoming_total_usd: number;
  outgoing_total_usd: number;
};

export type TopEntityStat = {
  entity_id: number;
  entity_name: string;
  amounts: Record<string, number>;
  total_usd: number;
};

export type TopTagStat = {
  tag_id: number;
  tag_name: string;
  amounts: Record<string, number>;
  total_usd: number;
};

export type EntityStatsBundle = {
  cached: boolean;
  balance_changes: EntityBalanceChangeByDay[];
  transactions_by_day: { day: string; transaction_count: number }[];
  money_flow_by_day: EntityMoneyFlowByDay[];
  top_incoming: TopEntityStat[];
  top_outgoing: TopEntityStat[];
  top_incoming_tags: TopTagStat[];
  top_outgoing_tags: TopTagStat[];
};

export type EntityStatsParams = {
  months?: number;
  limit?: number;
  signal?: AbortSignal;
};

export const getEntityStatsBundle = async (
  entityId: number,
  params: EntityStatsParams = {}
): Promise<EntityStatsBundle> => {
  const search = new URLSearchParams();
  if (params.months != null) search.set('months', String(params.months));
  if (params.limit != null) search.set('limit', String(params.limit));
  const query = search.toString();
  const url = `stats/entity/${entityId}${query ? `?${query}` : ''}`;
  return apiRequest<EntityStatsBundle>(url, { signal: params.signal });
};
